### ISSUE TITLE
High pass filter burn highlight

### ISSUE THREAD

**Issue by Pat-sz** [darktable-org/darktable#14070]
**Describe the bug/issue**
When I used the high pass filter, I selected a mask to apply only to the bird, but when I chose the 'overlay' blending mode, I saw that my highlight was burned and lost texture. It seems, if you see the histogram, that there is some highlight clipping.

**To Reproduce**

1. Go to 'Active High pass filter'
2. Active any one of fusion mode and select overlay
3. Transparency to 0
4. See the highlight burn on the photo

**Expected behavior**
Highlight should not be burned and clipped.

**Screenshots**
Without High pass filter
![image](https://user-images.githubusercontent.com/84800325/228852339-9aef342d-b94b-4edc-af2a-07927c65ada2.png)

With High pass filter active with overlay mode and 0% transparency
![image](https://user-images.githubusercontent.com/84800325/228906210-2232cecc-3c3e-4ccf-85b3-2f0e86a708f6.png)

**Platform**

* darktable version : 4.3.0+63~gc207f2ef3
* OS                : Windows family 11 (but dev mode active and run a Linux version of DT)
* Memory            : 16 Go
* Graphics card     : AMD Radeon RX 6500M
* Graphics driver   : AMD default
* OpenCL installed  : yes
* OpenCL activated  : yes
* Xorg              : unknow
* Desktop           : unknow
* GTK+              : unknow
* gcc               : unknow
* cflags            : unknow
* CMAKE_BUILD_TYPE  : unknow
* 

**Additional context**
I can reproduce on my DT 4.2.0 version on Windows, and on other photos that have a lot of highlighted detail.

**Comment by MStraeten** [comment]
in Lab and RGB screen color mode you‘re in a limited space, so there’s clipping if there’re values outside the 0-1 range. The rgb scene mode is able to deal with an unlimited range wider than 0-1

**Comment by Pat-sz** [comment]
@MStraeten Ok, but if the transparency is at 0%, the color of the photo should not change, right? And it's a blending mode. He seems me that, in the pipeline, the previous stage of the treatment should not be impacted by the filter.

**Comment by ralfbrown** [comment]
The blending modes perform a weighted combination of the module's input and output, with the opacity (uniform or drawn/parametric) controlling how much each contributes.  For example, in "normal" blend, the result is (output * opacity + input * (1-opacity)).  Most of the display-referred blend modes clip to 0...1 no matter what the opacity, generally because the math __breaks down__ if the input is not in that range.  You might have noticed that there is both a "normal" and a "normal bounded" mode - the bounded version clips while the other does not, but that's somewhat of an exception because the formula I gave above works properly regardless of inputs.

The formula for overlay blend is
$$L_{blend} = (1-opacity^2) * L_{in} + opacity^2 * (L_{in} > 0.5 ? 1 - (1 - 2*(L_{in}-0.5)) * (1 - L_{out}) : (2 * L_{in}) * L_{out})$$
which yields a negative result for $L_{in} > 1$ or ($L_{out} > 1$ and $L_{in} > 0.5$) due to the $(1 - 2*(L_{in}-0.5))$ and $(1 - L_{out})$ terms.  So $L_{in}$ and $L_{out}$ are clipped before applying the formula.

**Comment by Pat-sz** [comment]
Yes I understand, @ralfbrown, but in a photographer's logic, when you create a mask, it is to apply a transformation only under the mask and the other part must not be impacted, because of the transparency equal to 0. In my photo, there is a mask on the bird. I don't expect my highlights on the branch to be burned and clipped.

In this case, if this formula applies to another module, with overlay blending, I should observe the same clipping, but this is not the case in all Lab process modules like local contrast modules or color saturation.
In the High Pass module, I observe clipping in all blending modes, not specific in overlay. Overlay mode is just the normal blending mode to improve sharpness. That's the normal way to use it, I think.

Maybe this bug is due to the 0 in high pass module, is normally a neutral gray with 0,5 (RGB) value, but when you create a mask outside the value is the value of the enter layer is used. 

As I know what is the high pass formula, when I put my mask I see this:
![2023-03-31 11_04_59-darktable-bird](https://user-images.githubusercontent.com/84800325/229079701-cc11d083-e19f-492a-b5c3-e4d33bfe9195.png)

but I expected that because the zero is a neutral gray in high pass:
![2023-03-31 11_04_59-darktable-bird-example](https://user-images.githubusercontent.com/84800325/229079793-5a14f708-ff4a-4277-9834-1274494c40c2.jpg)

**Comment by MStraeten** [comment]
local contrast is in the display referred part of the pipeline (above filmic or sigmoid) while high pass is part of the scene referred pipeline. Filmic or sigmoid squeezes the scene referred range in a 0-1 display referred range.
If you want to have the same behavior for high pass as for local contrast you can move that module into the display referred part of the pipe. But then with all limits of the display referred workflow...

**Comment by Pat-sz** [comment]
That doesn't answer my question, @MStraeten. I want the high pass module to do what it needs to do. No change should produce gray pixels. Why outside the mask, where there should be gray information (which expects high-pass information), are there color pixels?
The high pass module produces abnormal results, but only in highlights and unmasked areas.
Am I so stupid that I'm mistaken in thinking that a high pass module should create a gray image?

**Comment by Pat-sz** [comment]
After a few tests to preserve my highlights, if I want to use the high pass module, I have to under-exposure by 0.7 EV, and at the end adjust the exposure with the level module and correct the saturation of the image which is far more prominent in highlights than before these changes. -0.7 EV correction: Clipping highlights are very important!

**Comment by Pat-sz** [comment]
I understand, that's not the problem. I think, to be consistent, if the reference value in the pass high module is neutral gray then 0% visibility should be translated as neutral gray. All arithmetic transformations must be performed in grayscale and translated into this color space. Otherwise, it makes no sense to use masks in all blending modes, and it leads to unexpected renders.

**Comment by Pat-sz** [comment]
I understand, that's not the problem. I think, to be consistent, if the reference value in the pass high module is neutral gray then 0% visibility should be translated as neutral gray. All arithmetic transformations must be performed in grayscale and translated into this color space. Otherwise, it makes no sense to use masks in all blending modes, and it leads to unexpected renders.

**Comment by Pat-sz** [comment]
I really want to trust you, but it's abstract. I need to see the input and output values to concretely understand how are the difference between data values and why it's necessary to clip it in highlights. And why I observe that in High pass module and not in other modules which work in Lab color space. If clipping is necessary here, why not in the other modules?

**Comment by MStraeten** [comment]
it’s worth to spend some time to get a basic understanding of the darktable pixelpipe: https://darktable-org.github.io/dtdocs/en/darkroom/pixelpipe/ 
As I wrote earlier: high pass is applied in the part of the pipe where there’s a 0-unlimited range of tonal values. Local contrast is in the part of the pipe where there’s just a 0-1 range.
If you‘re using blend modes, that just supports a 1-0 range these aren’t able to cope with those information in the 1-unlimited range.
At the border between scene referred and display referred parts of the pipe filmic-rgb(or sigmoid) squeezes the 0-unlimited range into a 0-1 range in a way you can decide how much of that initial range you want to have preserved. This tonemapping must be done explicitly so if you want to use display referred blend for a module modes you should move it after the tonemapping step if you don’t just preserve initial 0-1 part of the 0-unlimited range. 

the middle column in https://discuss.pixls.us/t/is-the-display-referred-workflow-a-destructive-workflow-and-the-scene-referred-non-destructive/29608/5 illustrates, what’s happening if using display referred blend modes in teh scene referred part
